### ISSUE TITLE
Synapse does not apply edits to events other than `m.room.message`

### ISSUE THREAD

**Issue by richvdh** [matrix-org/synapse#12793]
MSC2676 [says](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#server-behaviour):

> Whenever a homeserver would return an event via the Client-Server API, it should check for any valid, applicable `m.replace` event, and if one is found, it should first modify the content of the original event according to the `m.new_content` of the most recent edit.

It also [gives](https://github.com/uhoreg/matrix-doc/blob/aggregations-edits/proposals/2676-message-editing.md#validity-of-message-edit-events) a comprehensive list of what comprises a "valid" edit event.

Synapse does this, but only where the events are of type `m.room.message`. This is incorrect: it should also do so for other event types. Examples include polls ([`org.matrix.msc3381.poll.start`](https://github.com/matrix-org/matrix-spec-proposals/pull/3381)), which can be edited via the Element UI, and stickers ([`m.sticker`](https://spec.matrix.org/v1.2/client-server-api/#sticker-messages)), which cannot be edited via the UI but can via the API.

This is closely related to #12503, but that concerns aggregations, whereas this issue is about updating the `content` of the original event.

**Comment by alariej** [comment]
I came across this issue while doing tests on matrix.org. As far as I can tell, Synapse doesn't even seem to "modify the content of the original event according to the m.new_content of the most recent edit.", even for `m.room.message` events. The original message content does not get modified after one or even several edits. Below are the message objects from a test I just did on matrix.org. The message was created and edited in a fresh non-encrypted private room using Element Web. After editing the message, I logged out, refreshed the browser, and logged back in to make sure the events were fetched again from the server.

Original message, before and after editing:

```
content:
  body: "hi"
  msgtype: "m.text"
  org.matrix.msc1767.text: "hi"
event_id: "$PbrLqfPm7MBDHLZd5gVREn3VA8UyFc8qb9220xiKYlw"
origin_server_ts: 1665067355108
sender: "@xyz:matrix.org"
type: "m.room.message"
```

Message created by editing:

```
content: 
  body: " * hi hi"
  msgtype: "m.text"
  org.matrix.msc1767.text: " * hi hi"
  m.new_content: 
    body: "hi hi"
    msgtype: "m.text"
    org.matrix.msc1767.text: "hi hi"
  m.relates_to: 
    event_id: "$PbrLqfPm7MBDHLZd5gVREn3VA8UyFc8qb9220xiKYlw"
    rel_type: "m.replace"
event_id: "$rOEXMZKBNp8iYoBcSvdOFIE7bAqVVpJGLhc68ENPLw8"
origin_server_ts: 1665067377760
sender: "@xyz:matrix.org"
type: "m.room.message"
```
I guess this would not get picked up by users as Element will anyway display the edited content of messages, but I'm wondering if something has changed in the protocol or implementation since this issue was opened. Or am I missing something completely obvious here?

**Comment by alariej** [comment]
Via the sync API.

**Comment by alariej** [comment]
Oh right, I had missed the point with the "limited". Not quite sure about the logic of it, but in any case, thanks for the info!

**Comment by clokep** [comment]
See #14252.

**Comment by clokep** [comment]
Hm actually I think this was fixed by #15193 to no longer apply edits.
